raiden/network/transport/matrix.py

```diff
@@ -338,7 +356,7 @@ def _user_id(self) -> Optional[str]:
         return getattr(self, '_client', None) and getattr(self._client, 'user_id', None)
 
     @property
-    @cachedmethod(_cachegetter('__log_cache', dict), key=attrgetter('_user_id'))
+    @cachedmethod(_factorygetter('__log_cache', dict), key=attrgetter('_user_id'))
```

Comment:
We could also use the `__init__` method mentioned above to declare log cache and use the `cachedmethod` without the `factorygetter`. Similar to what's implemented in the [example here](https://cachetools.readthedocs.io/en/latest/#cachetools.cachedmethod)

Reply:
Yes, we could, but I thought this decorator/declarative implementation was better

Reply:
Also, I can't see how subclassing whole 3rd-party classes only to add a lock to it is a better pattern than the one I implemented.

Reply:
So, from what I understood the cache here makes the property useless. On the first call the logger will be created and cached, on all subsequent calls the same loggers will be returned. I looked at the documentation for `cachedmethod` and it doesn't state explicit what is the TTL, but even with a TTL this doesn't make sense because until the cache expires the binding will not happen.

Please, avoid using caches for everything, I doubt the performance hit of this was measured.

Reply:
@hackaugusto please, look closer:
`_factorygetter` is just as `attrgetter`, it'll return the cache (in this case, `dict`) **per instance**. Besides that, the `key` parameter will cache the logger **by `_user_id`**, meaning if it changes **in that instance** (which isn't supposed to happen anyway), it'll get another, properly bound logger. Yes, this cache is useful to prevent the property calling `log.bind` on **every** log call.

Reply:
Humm, I see. I definitely missed `key=attrgetter('_user_id'))`, which to me definitely points to the unreadability of the code.

> Yes, this cache is useful to prevent the property calling log.bind on every log call.

Yes, did you measure the impact of that before introducing the cache?

Reply:
> definitely points to the unreadability of the code.

please, open an issue on `cachetools` and python upstream for `attrgetter` then, it's documented there.

> Yes, did you measure the impact of that before introducing the cache?

No, previously we even had `if logger.isEnabledFor` calls, I took the freedom to avoid these calls on every log call to cache it, as it definitelly doesn't have the need to be called everytime, only when `_user_id` changes.